File: leetcode_python/Dynamic_Programming/binary-trees-with-factors.py
# V0 

# V1 
# http://bookshadow.com/weblog/2018/04/22/leetcode-binary-trees-with-factors/
"""

DP def
    (arr is SORTED first, so every factor is processed before its multiple)

    dp[a]: number of binary trees whose ROOT value is a

           (each non-leaf node's value == product of its 2 children)

DP eq

     dp[a] = 1 + sum( dp[m] * dp[a/m] * (1 if m == a/m else 2) )

             for every m in arr with m * m <= a and a % m == 0


    -> e.g.
         the leading 1     = the single-node tree (just `a` itself)
         the (1 if .. 2)   = left/right children can be swapped when m != a/m

     ans = sum(dp.values()) % (10^9 + 7)

"""
# time = O(n^2)
# space = O(n)
import collections
class Solution(object):
    def numFactoredBinaryTrees(self, A):
        """
        :type A: List[int]
        :rtype: int
        """
        A.sort()
        dp = collections.defaultdict(int)
        MOD = 10 ** 9 + 7
        for i, a in enumerate(A):
            num = 0
            for j in range(i):
                m = A[j]
                if m * m > a: break
                if a % m: continue
                n = a / m
                num = (num + dp[m] * dp[n] * (1 + (m != n))) % MOD
            dp[a] = num + 1
        return sum(dp.values()) % MOD

# V1'
# https://www.jiuzhang.com/solution/binary-trees-with-factors/#tag-highlight-lang-python
"""

DP def
    (arr is SORTED first, so every factor is processed before its multiple)

    dp[a]: number of binary trees whose ROOT value is a

           (each non-leaf node's value == product of its 2 children)

DP eq

     dp[a] = 1 + sum( dp[m] * dp[a/m] * (1 if m == a/m else 2) )

             for every m in arr with m * m <= a and a % m == 0


    -> e.g.
         the leading 1     = the single-node tree (just `a` itself)
         the (1 if .. 2)   = left/right children can be swapped when m != a/m

     ans = sum(dp.values()) % (10^9 + 7)

"""
# time = O(n^2)
# space = O(n)
class Solution:
    def numFactoredBinaryTrees(self, A):
        A.sort()
        dp = {}
        for i in range(len(A)):
            dp[A[i]] = 1
            for j in range(i):
                if A[i] % A[j] == 0 and A[i] / A[j] in dp:
                    dp[A[i]] += dp[A[j]] * dp[A[i] / A[j]]
        return sum(dp.values()) % (10**9 + 7)

# V2
"""

DP def
    (arr is SORTED first, so every factor is processed before its multiple)

    dp[a]: number of binary trees whose ROOT value is a

           (each non-leaf node's value == product of its 2 children)

DP eq

     dp[a] = 1 + sum( dp[m] * dp[a/m] * (1 if m == a/m else 2) )

             for every m in arr with m * m <= a and a % m == 0


    -> e.g.
         the leading 1     = the single-node tree (just `a` itself)
         the (1 if .. 2)   = left/right children can be swapped when m != a/m

     ans = sum(dp.values()) % (10^9 + 7)

"""
# time = O(n^2)
# space = O(n)
class Solution(object):
    def numFactoredBinaryTrees(self, A):
        """
        :type A: List[int]
        :rtype: int
        """
        M = 10**9 + 7
        A.sort()
        dp = {}
        for i in range(len(A)):
            dp[A[i]] = 1
            for j in range(i):
                if A[i] % A[j] == 0 and A[i] // A[j] in dp:
                    dp[A[i]] += dp[A[j]] * dp[A[i] // A[j]]
                    dp[A[i]] %= M
        return sum(dp.values()) % M

File: leetcode_java/src/main/java/LeetCodeJava/DynamicProgramming/BinaryTreesWithFactors.java
package LeetCodeJava.DynamicProgramming;

// https://leetcode.com/problems/binary-trees-with-factors/

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

/**
 *  823. Binary Trees With Factors
 *  Medium
 *
 *  Given an array of unique integers, arr, where each integer arr[i] is
 *  strictly greater than 1.
 *
 *  We make a binary tree using these integers, and each number may be used for
 *  any number of times. Each non-leaf node's value should be equal to the
 *  product of the values of its children.
 *
 *  Return the number of binary trees we can make. The answer may be too large
 *  so return the answer modulo 10^9 + 7.
 *
 *  Example 1:
 *    Input: arr = [2,4]
 *    Output: 3
 *    Explanation: We can make these trees: [2], [4], [4, 2, 2]
 *
 *  Example 2:
 *    Input: arr = [2,4,5,10]
 *    Output: 7
 *    Explanation: We can make these trees:
 *                 [2], [4], [5], [10], [4, 2, 2], [10, 2, 5], [10, 5, 2].
 *
 *  Constraints:
 *    - 1 <= arr.length <= 1000
 *    - 2 <= arr[i] <= 10^9
 *    - All the values of arr are unique.
 */
public class BinaryTreesWithFactors {

    private static final int MOD = 1_000_000_007;

    // V0
    // IDEA: sort ascending, dp[i] = # trees rooted at arr[i];
    //       every child pair (j, arr[i]/arr[j]) must be smaller, so it is already computed
    /**
     * time = O(N^2)
     * space = O(N)
     */
    public int numFactoredBinaryTrees(int[] arr) {
        int n = arr.length;
        int[] a = Arrays.copyOf(arr, n);
        Arrays.sort(a);

        Map<Integer, Integer> idx = new HashMap<>();
        for (int i = 0; i < n; i++) {
            idx.put(a[i], i);
        }

        long[] dp = new long[n];
        long res = 0;
        for (int i = 0; i < n; i++) {
            dp[i] = 1;   // the single-node tree
            for (int j = 0; j < i; j++) {
                if (a[i] % a[j] != 0) {
                    continue;
                }
                int other = a[i] / a[j];
                Integer k = idx.get(other);
                if (k != null) {
                    dp[i] = (dp[i] + dp[j] * dp[k]) % MOD;
                }
            }
            res = (res + dp[i]) % MOD;
        }
        return (int) res;
    }
}
